Buffer python warning to avoid deadlocks

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27154 Handle warning in torchscript
* #26614 add pybind version of HANDLE_TH_ERRORS
* **#26613 Buffer python warning to avoid deadlocks**
* #28283 Fix clang-tidy errors in csrc/Module.cpp

Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

albanD committed:
Buffer python warning to avoid deadlocks

[ghstack-poisoned]

albanD committed:
Update on "[WIP] Buffer python warning to avoid deadlocks"




[ghstack-poisoned]

ezyang commented:
You need to rebase to master to fix build failures

ezyang commented on torch/csrc/Exceptions.h:
```diff
@@ -138,4 +157,27 @@ struct ValueError : public PyTorchError {
   }
 };
 
+// ATen warning handler for Python
```
I see this code was moved. Any substantive changes inside?

albanD replied:
Yes only the comment was moved. The rest of the code is new !

ezyang commented on torch/csrc/Exceptions.h:
```diff
@@ -138,4 +157,27 @@ struct ValueError : public PyTorchError {
+    const c10::SourceLocation& source_location,
+    const std::string& msg);
+
+  static warning_buffer_t warning_buffer;
```
This struct is all static members. Any reason for it to be a struct? A namespace would have worked equally as well. You can move the `warning_buffer` declaration to be private to `torch/csrc/Exceptions.cpp` as it doesn't seem to be used anywhere else.

ezyang replied:
Is `warning_buffer` thread safe? I don't think `std::queue` is thread safe. So what lock protects it?

ezyang commented on torch/csrc/Exceptions.h:
```diff
@@ -138,4 +157,27 @@ struct ValueError : public PyTorchError {
+public:
+/// See NOTE [ Conversion Cpp Python Warning ] for noexcept justification
+  EnforceWarningBuffer() noexcept(true);
+  ~EnforceWarningBuffer() noexcept(false);
```
Wow, TIL this is allowed hahaha

albanD replied:
This is my attempt to make sure that we don't get killed by throwing two exceptions.

ezyang commented on torch/csrc/Exceptions.h:
```diff
@@ -2,38 +2,57 @@
+
+/// NOTE [ Conversion Cpp Python Warning ]
+/// Python warning semantic is different from the cpp one in that
+/// they can raise errors. To handle this case which requires
```
I'm not sure this first sentence makes sense to anyone who didn't participate in the design discussion. An example would make it a lot clearer, I think.

ezyang replied:
Basically, walk the reader through the case when a warning handler raises an error, while the function itself is throwing an exception. (That diagram we drew, but in comments.)

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

ezyang commented on c10/util/Exception.cpp:
```diff
@@ -64,16 +64,20 @@ void Error::AppendMessage(const std::string& new_msg) {
 
 void Warning::warn(SourceLocation source_location, std::string msg) {
-  warning_handler_(source_location, msg.c_str());
+  warning_handler_(source_location, msg);
```
nit: `std::move(msg)` here

ezyang replied:
Actually, it looks like the warning handler takes this by const reference. In that case, warn() should also take by const reference...

ezyang commented on torch/csrc/Exceptions.cpp:
```diff
@@ -137,5 +137,73 @@ ValueError::ValueError(const char *format, ...) {
+      // A python error happened after the warning
+      // Simply handle with the cpp handler
+      while(warning_buffer.size() > 0) {
+        auto warning = warning_buffer.front();
```
`auto warning = std::move(warning_buffer.front());` here is better, it avoids an unnecessary copy assignment

ezyang replied:
But tbh, C++ `queue` is really bad and there isn't really a good reason to use it here: you can just stuff everything on a vector, then process the vector front to back and reset() it to empty when you're all done. (You don't need the ability to incrementally remove items; you only need to remove them all at once in the end.)

ezyang commented on torch/csrc/Exceptions.cpp:
```diff
@@ -137,5 +137,73 @@ ValueError::ValueError(const char *format, ...) {
+        warning_buffer.pop();
+        auto source_location = warning.first;
+        auto msg = processErrorMsg(warning.second);
+        c10::Warning::warn(source_location, msg);
```
The use of `warn` here suggests to me that there is no good reason not to pass the `std::string` in by value all the way, as the caller never needs to retain `msg` in the end.

ezyang commented:
Looks good. Just some C++ bits; I also wanna know what the locking strategy is.

ezyang commented:
Don't forget to request re-review (click the little circle arrow icon next to the reviewer) when you want rereview

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD commented:
@ezyang  Clang-tidy errors are unrelated to my changes, can I ignore it?

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

ezyang commented:
Yes, please don't fix clang tidy in this PR.

On second review, I'm also feeling like there is a missing comment saying *why* we are going through all of this rigamarole in the first place. Obviously we have had the design discussion in person, but none of that design discussion has been replicated, neither in the code, or the PR body.

ezyang commented on torch/csrc/Exceptions.cpp:
```diff
@@ -137,5 +137,81 @@ ValueError::ValueError(const char *format, ...) {
+  std::unique_lock<std::mutex> lock(warning_buffer_mutex);
+
+  if(warning_buffer.size() > 0) {
+    AutoGIL gil;
```
Can `py_warning_handler` ever be called when we have the GIL? In that case, there's potential for deadlock here, right? One thread has the GIL and is waiting on warning buffer mutex, the other thread has warning buffer mutex and is waiting on the GIL.

ezyang reviewed (changes requested):
deadlock

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

ezyang commented on torch/csrc/Exceptions.cpp:
```diff
@@ -137,5 +137,89 @@ ValueError::ValueError(const char *format, ...) {
+  bool has_warnings;
+  {
+    std::unique_lock<std::mutex> lock(warning_buffer_mutex);
+    has_warnings = warning_buffer.size() > 0;
```
Hmmm. While it is certainly true that you avoid the deadlock this way, taking out a lock to test size, and then releasing the lock, and then taking the lock out later is definitely a concurrency antipattern. What if another thread modifies `warning_buffer` in-between the time you check and then when you actually run?

But this raises further questions in my mind: why is the warning buffer global (as opposed to something thread local?)

albanD replied:
Because it needs to be accessed from the static `at::warn()` function.
Is there a way to bind to bind to call `static void set_warning_handler(handler_t handler);` with a handler which is a function associated with this class instance? For example if I had a `EnforceWarningBuffer::warn_handler()` (non static function).

albanD replied:
It is not thread local mainly because that would force us to duplicate all the logic that currently exists for errors to move them from one thread to the other whenever multithreaded is used.

ezyang replied:
Well, if the warning handler takes a straight up function pointer there's nothing you can do, but if you change it to accept something more interesting you would be in better shape. Like it takes a pointer with a virtual method on it to handle.

The justification for not TLS feels flimsy to me, though I understand from a local perspective why it would be nicer to avoid having to solve the TLS transfer across threads problem. In particular, the reason why it feels dodgy to me is because it means one thread can queue warnings, and then they magically show up from another thread when returning to Python. Warnings will literally escape the scopes they were supposed to live in...

albanD replied:
After discussion. I'll update the PR to detect when such "escape" can happen. And warn the user in that case.

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD commented:
After more issues and more discussion. We're going to move this handler to a thread local variable.

Commit:
Update on "[WIP] Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

Commit:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D17692505](https://our.internmc.facebook.com/intern/diff/D17692505)

[ghstack-poisoned]

ezyang commented on c10/util/Exception.h:
```diff
@@ -82,26 +82,31 @@ class C10_API Error : public std::exception {
+/// it is not needed.
+C10_API void set_warning_handler(WarningHandler* handler) noexcept(true);
+/// Gets the global warning handler.
+C10_API WarningHandler* get_warning_handler() noexcept(true);
```
Nit: expose this functionality as an RAII guard so people don't forget to reset the warning handler when control exits, e.g., via an exception

ezyang replied:
I guess PyWarningHandler brings its own RAII thing.

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

ezyang commented on docs/cpp/source/Doxyfile:
```diff
@@ -70,6 +70,8 @@ RECURSIVE              = YES
 EXCLUDE = ../../../torch/csrc/api/include/torch/nn/pimpl-inl.h \
           ../../../torch/csrc/api/include/torch/detail
+# Increase the max node size for our large files
+DOT_GRAPH_MAX_NODES    = 100
```
don't put this in an unrelated diff, this is what stacks are for!! Cross referencing with  https://github.com/pytorch/pytorch/pull/29041

albanD replied:
This is actually related. the cpp docs does not build for Exception.h without this.
See build failures on the previous commit here: https://github.com/pytorch/pytorch/runs/284960182

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]

albanD committed:
Update on "Buffer python warning to avoid deadlocks"


Differential Revision: [D18249633](https://our.internmc.facebook.com/intern/diff/D18249633)

[ghstack-poisoned]
